standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlInsertPart.java

```diff
@@ -0,0 +1,804 @@
+    int rowCount = 0;
+    for (MPartition part : partIdToPartition.values()) {
+      rowCount += part.getParameters().size();
+    }
+    if (rowCount == 0) {
+      return;
+    }
+    String columns = "(\"PART_ID\",\"PARAM_KEY\",\"PARAM_VALUE\")";
+    int columnCount = 3;
+    BatchExecutionContext batchExecutionContext = new BatchExecutionContext() {
+      final Iterator<Map.Entry<Long, MPartition>> partIt = partIdToPartition.entrySet().iterator();
+      Map.Entry<Long, MPartition> partEntry = partIt.next();
+      Iterator<Map.Entry<String, String>> it = partEntry.getValue().getParameters().entrySet().iterator();
+      @Override
+      public void execute(String batchQueryText, int batchRowCount, int batchParamCount) throws MetaException {
+        Object[] params = new Object[batchParamCount];
+        int index = 0;
+        int paramIndex = 0;
+        do {
+          while (index < batchRowCount && it.hasNext()) {
+            Map.Entry<String, String> entry = it.next();
+            params[paramIndex++] = partEntry.getKey();
+            params[paramIndex++] = entry.getKey();
+            params[paramIndex++] = entry.getValue();
+            index++;
+          }
+          if (index < batchRowCount) {
+            partEntry = partIt.next();
+            it = partEntry.getValue().getParameters().entrySet().iterator();
+          }
+        } while (index < batchRowCount);
+        executeQuery(batchQueryText, params);
+      }
+    };
+    insertInBatch("\"PARTITION_PARAMS\"", columns, columnCount, rowCount, batchExecutionContext);
+  }
+
+  private void insertPartitionKeyValInBatch(Map<Long, MPartition> partIdToPartition) throws MetaException {
+    int rowCount = 0;
+    for (MPartition part : partIdToPartition.values()) {
+      rowCount += part.getValues().size();
+    }
+    if (rowCount == 0) {
+      return;
+    }
+    String columns = "(\"PART_ID\",\"PART_KEY_VAL\",\"INTEGER_IDX\")";
+    int columnCount = 3;
+    BatchExecutionContext batchExecutionContext = new BatchExecutionContext() {
+      int colIndex = 0;
+      final Iterator<Map.Entry<Long, MPartition>> partIt = partIdToPartition.entrySet().iterator();
+      Map.Entry<Long, MPartition> partEntry = partIt.next();
+      Iterator<String> it = partEntry.getValue().getValues().iterator();
+      @Override
+      public void execute(String batchQueryText, int batchRowCount, int batchParamCount) throws MetaException {
+        Object[] params = new Object[batchParamCount];
+        int index = 0;
+        int paramIndex = 0;
+        do {
+          while (index < batchRowCount && it.hasNext()) {
+            params[paramIndex++] = partEntry.getKey();
+            params[paramIndex++] = it.next();
+            params[paramIndex++] = colIndex++;
+            index++;
+          }
+          if (index < batchRowCount) {
+            colIndex = 0;
+            partEntry = partIt.next();
```

Comment:
same as `sdIt` above

Reply:
same reason as above. Have added comment now to make it clear.